app/models/user.rb

```diff
@@ -4,10 +4,36 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
-  has_one :homepage, dependent: :destroy
-  after_create -> { self.create_homepage }
+  # Virtual attribute for authenticating by either username or email
+  # This is in addition to a real persisted field like 'username'
+  attr_accessor :login
 
+  validates_format_of :username, with: /^[a-zA-Z0-9_\.]*$/, multiline: true
+  validates :username,
+            presence: true,
+            uniqueness: { case_sensitive: false }
+
+  has_one :homepage, dependent: :destroy
   has_many :lists
 
+  before_save { self.email.downcase! if self.email }
+  after_create -> { self.create_homepage }
+
   acts_as_voter
+
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions.to_h).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
+    elsif conditions.has_key?(:username) || conditions.has_key?(:email)
+      conditions[:email].downcase! if conditions[:email]
+      conditions[:username].downcase! if conditions[:username]
```

Comment:
This introduces the problem that I can save a `User` with a case sensitive username and then have it not return correctly if I give `#find_for_database_authentication` the proper username.

```
irb(main):003:0> u
=> #<User ... username: "USeR">
irb(main):004:0> User.find_for_database_authentication username: 'USeR'
  User Load (0.8ms)  SELECT  "users".* FROM "users" WHERE "users"."username" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["username", "user"], ["LIMIT", 1]]
=> nil
```